Security.md

```diff
@@ -108,13 +108,76 @@ in-order execution and [post-MVP atomic memory primitives](PostMVP.md#threads).
 modules. In the future, additional protections may be provided by runtimes or
 the toolchain, such as code diversification or memory randomization (similar to
-[address space layout randomization][] (ASLR)), [bounded pointers][] ("fat"
-pointers), or finer-grained control-flow integrity.
+[address space layout randomization][] (ASLR)), or [bounded pointers][] ("fat"
+pointers).
+
+### Control-Flow Integrity
+The effectiveness of control-flow integrity can be measured based on its
+completeness. Generally, there are three types of control-flow transitions that
+need to be protected:
+  1. Direct function calls,
+  2. Indirect function calls,
+  3. Returns.
+
+Together, (1) and (2) are commonly referred to as "forward-edge", since they
+correspond to forward edges in a directed control-flow graph. Likewise (3) is
+commonly referred to as "back-edge", since it corresponds to back edges in
+a directed control-flow graph.
+
+Typically, this is implemented using runtime instrumentation. During
+compilation, the compiler generates an expected control flow graph of program
+execution, and inserts runtime instrumentation at each call site to verify that
+the transition is safe. Sets of expected call targets are constructed from the
+set of all possible call targets in the program, unique identifiers are assigned
+to each set, and the instrumentation checks whether the current call target is
+a member of the expected call target set. If this check succeeds, then the
+original call is allowed to proceed, otherwise a failure handler is executed,
+which typically terminates the program.
+
+In WebAssembly, the execution semantics implicitly guarantee the safety of (1)
+through usage of explicit function section indexes, and (3) through a shadow
+stack. Additionally, the type signature of indirect function calls is already
+checked at runtime, effectively implementing coarse-grained type-based control-
+flow integrity for (2). All of this is achieved without explicit runtime
+instrumentation in the module. However, as discussed
+[previously](#memory-safety), this protection does not prevent code reuse
+attacks with function-level granularity against indirect calls. As a result,
+the WebAssembly also supports fine-grained control-flow integrity for (2). This
+leverages built-in support for [forward-edge control-flow integrity][] in the
+Clang/LLVM compiler suite, which has been extended to WebAssembly.
+
+Support for fine-grained control-flow integrity in WebAssembly operates as
+follows. First, the Clang frontend emits type signatures for each function
+and each call site in the LLVM IR. This is used to partition all accessible
+functions in the module into separate expected call target sets based on type.
+Next, instrumentation is inserted at each indirect call site to verify that the
+computed index has the correct type. Currently, a transformation pass
+(`LowerTypeTests`) performs this operation, by using a monotonic counter to
```

Comment:
Going into the specifics of Clang's lowering code seems out of place. I think the extent of what we should say is that more fine-grained CFI is possible for producers, and maybe note something more general like "this is possible by organizing indirect function tables and adding range checks before calls (a technique currently utilized by clang)", etc.